Tutorial - Angular surveys (#7596)
PostHog/posthog.com: contents/tutorials/angular-surveys.md

import EventsLight from '../images/tutorials/angular-surveys/events-light.png'
import EventsDark from '../images/tutorials/angular-surveys/events-dark.png'
import ImgSurveyResultsLight from '../images/tutorials/angular-surveys/survey-results-light.png'
import ImgSurveyResultsDark from '../images/tutorials/angular-surveys/survey-results-dark.png'
import ImgSurveyTemplatesLight from '../images/tutorials/angular-surveys/survey-templates-light.png'
import ImgSurveyTemplatesDark from '../images/tutorials/angular-surveys/survey-templates-dark.png'

[Surveys](/docs/surveys) are a great way to get feedback from your users. In this guide, we show you how to add a survey to your Angular.js app.

We'll create a basic Angular app, add PostHog, create a survey, and then show you how to display the survey in the app and get responses.

## 1. Create a Vue app

First, ensure [Node.js is installed](https://nodejs.dev/en/learn/how-to-install-nodejs/) (version 14.20.0 or newer). Then, install the Angular CLI and create a new Angular app:

```bash
npm install -g @angular/cli
ng new angular-surveys
```

Select `CSS` as your stylesheet and `No` for server side rendering and static site generation.

> Next, Replace the code in `src/app/app.component.html` with a simple heading and button:

Author: ```suggestion
Next, Replace the code in `src/app/app.component.html` with a simple heading:
```

```html file=app.component.html
<div id="app">
  <h1>Angular Surveys</h1>
</div>
```

Run `ng serve` and navigate to http://localhost:4200 to see your app in action.

![Basic Angular app](../images/tutorials/angular-ab-tests/basic-app.png)

## 2. Add PostHog

We use PostHog to create and control our survey as well as monitor results. If you don't have a PostHog instance, you can [sign up for free here](https://us.posthog.com/signup). 

To start, install the [JavaScript web SDK](/docs/libraries/js):

```bash
npm i posthog-js
```

In `src/main.ts`, initialize PostHog using your project API key and instance address. You can get both in your [project settings](https://us.posthog.com/project/settings).  

```ts file=main.ts
import { bootstrapApplication } from '@angular/platform-browser';
import { appConfig } from './app/app.config';
import { AppComponent } from './app/app.component';
import posthog from 'posthog-js'

posthog.init(
  '<ph_project_api_key>',
  {
    api_host:'<ph_instance_address>'
  }
)

bootstrapApplication(AppComponent, appConfig)
  .catch((err) => console.error(err));
```

> Once you’ve done this, reload your app and click the button a few times. You should see events appearing in the [PostHog events explorer](https://us.posthog.com/events).

Author: ```suggestion
Once you’ve done this, reload your app. You should see events appearing in the [PostHog events explorer](https://us.posthog.com/events).
```

### Option 1: Use PostHog's prebuilt survey UI

This is the simplest option. PostHog has a variety of [survey templates](/templates?filter=type&value=survey) to choose from, and handles all the display logic and response capture for you. You can also customize the questions, branding, and targeting as needed – see our [survey docs](/docs/surveys/creating-surveys) for more details on how to do so.
 
To create a survey with a prebuilt UI, go to the [surveys tab](https://us.posthog.com/surveys) in PostHog and click "New survey". 

<ProductScreenshot
  imageLight={ImgSurveyTemplatesLight} 
  imageDark={ImgSurveyTemplatesDark} 
  alt="PostHog survey templates" 
  classes="rounded"
/>

Select any template, or you can create your own by clicking "Create blank survey". Then, configure your survey with the following details:

1. Ensure `Presentation` is set to **Popover**.
2. Set the targeting to `All users`.
3. Use the default values for everything else.

Then, click "Save as draft" and then "Launch". Your survey is now live and you should see it in your app. After submitting responses, you can [view results in PostHog](#4-view-results).

> ![Popover survey in app](../images/tutorials/vue-surveys/popover-survey.png)

Author: ```suggestion
![Popover survey in app](../images/tutorials/angular-surveys/popover-survey.png)
```

@Component({
  selector: 'app-custom-survey',
  standalone: true,
  imports: [CommonModule], 
  templateUrl: './custom-survey.component.html',
  styleUrls: ['./custom-survey.component.css']
})
export class CustomSurveyComponent {
  @Input() title!: string;
  @Output() onDismiss = new EventEmitter<void>();
  @Output() onSubmit = new EventEmitter<number>();

  selectedValue: number | undefined = undefined;
  numbers = Array.from({ length: 10 }, (_, i) => i + 1);

  handleSelect(value: number) {
    this.selectedValue = value;
  }

  emitDismiss() {
    this.onDismiss.emit();
  }

  emitSubmit() {
    this.onSubmit.emit(this.selectedValue);
  }
}
```

In `custom-survey.component.css`:

```css file=custom-survey.component.css
.survey {
  position: fixed;
  bottom: 20px;
  right: 20px;
  width: 400px;
  padding: 20px;
  background-color: #ffffff;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.2);
  border-radius: 5px;
  z-index: 1000;
}

.button {
  margin: 5px;
  padding: 5px;
}
```

Now we integrate our new `CustomSurveyComponent` into our `AppComponent`. First, `<app-custom-survey>` to `app.component.html`:

```html file=app.component.html
<div id="app">
  <h1>Angular Surveys</h1>

>       <app-custom-survey

Reviewer: ```suggestion
    <app-custom-survey
```

      *ngIf="showSurvey"
      [title]="surveyTitle"
      (onDismiss)="handleDismiss()"
      (onSubmit)="handleSubmit($event)"
    />
</div>
```

Then, in `app.component.ts`, import `CustomSurveyComponent` and define the methods:

```typescript file=app.component.ts
import { Component } from '@angular/core';

> import { RouterOutlet } from '@angular/router';

Author: ```suggestion
import { CommonModule } from '@angular/common';
import { RouterOutlet } from '@angular/router';
```

> import { CommonModule } from '@angular/common';
> import { RouterOutlet } from '@angular/router';

Reviewer: RouterOutlet is there by default, so just swapping order
```suggestion
import { RouterOutlet } from '@angular/router';
import { CommonModule } from '@angular/common';
```

@Component({
  selector: 'app-root',
  standalone: true,

>   imports: [RouterOutlet, CustomSurveyComponent], 

Author: ```suggestion
  imports: [RouterOutlet, CustomSurveyComponent, CommonModule], 
```

Reviewer: Definitely need this 😄

  imports: [RouterOutlet, CustomSurveyComponent], 
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent {
  title = 'angular-surveys';
  surveyTitle = 'Survey title';
  showSurvey = true;

  handleDismiss() {
    this.showSurvey = false;
  }

  handleSubmit(value: number) {
    console.log("Submitted value:", value);
    this.showSurvey = false;
  }
}
```

This shows a survey popup every time you visit your app's homepage.

![Custom survey UI](../images/tutorials/angular-surveys/sample-survey-ui.png)

#### 2. Fetch the survey from PostHog

PostHog keeps track of all active surveys for a user (this is especially helpful if you set up [custom targeting options](/docs/surveys/creating-surveys#targeting)). 

> To fetch the active surveys, we use `posthog.getActiveMatchingSurveys()`. This returns a surveys object that looks like this:

Reviewer: ```suggestion
To fetch the active surveys, we use `posthog.getActiveMatchingSurveys()`. This returns an array of survey objects that looks like this:
```

```JSON
[
   {
     "id": "018cfcd5-107e-0000-49a1-8e7c6b825947",
     "name": "Net promoter score (NPS) API Survey",
     "description": "",
     "type": "api",
     "linked_flag_key": null,
     "targeting_flag_key": null,
     "questions": [
       {
         "type": "rating",
         "scale": 10,
         "display": "number",
         "question": "How likely are you to recommend us to a friend?",
         "description": "",
         "lowerBoundLabel": "Unlikely",
         "upperBoundLabel": "Very likely"
       }
     ],
     "conditions": null,
     "start_date": "2024-01-12T08:41:20.614000Z",
     "end_date": null
   }
]
```

> To fetch this object and integrate it with your survey UI, update your code in `app.component.ts`:

Reviewer: ```suggestion
To fetch this array and integrate it with your survey UI, update your code in `app.component.ts`:
```

```typescript file=app.component.ts
import { Component, ChangeDetectorRef } from '@angular/core';
import posthog from 'posthog-js'
// rest of your imports

@Component({
  // your existing config
})
export class AppComponent implements OnInit {
  title = 'angular-surveys';

>   showSurvey = true;
>   surveyTitle = 'Survey title';

Reviewer: Flipping to match order of last use
```suggestion
  surveyTitle = 'Survey title';
  showSurvey = true;
```

  constructor(private changeDetector: ChangeDetectorRef) {}

  ngOnInit(): void {
    this.fetchActiveSurveys();
  }

  private fetchActiveSurveys(): void {
    posthog.getActiveMatchingSurveys((surveys) => {
      if (surveys.length > 0) {
        const survey = surveys[0];
        this.surveyID = survey.id;
        this.surveyTitle = survey.questions[0].question;
        this.changeDetector.detectChanges();
      }
    });
  }

  handleDismiss(): void {
    // existing code
  }

  handleSubmit(value: number): void {
    // existing code
  }
}
```

#### 3. Add the logic for displaying and hiding it.

We want to make sure we don't show the survey again to users who have either submitted or dismissed it. We use [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) to store this data and use it to check whether to show the survey or not.

```typescript file=app.component.ts
// rest of your imports and code

export class AppComponent {
  title = 'angular-surveys';
  showSurvey = false; // updated to false

  // rest of code

  private fetchActiveSurveys(): void {
    posthog.getActiveMatchingSurveys((surveys) => {

>       console.log(surveys)

Author: ```suggestion
```

      if (surveys.length > 0) {
        const survey = surveys[0];
        this.surveyID = survey.id;
        this.surveyTitle = survey.questions[0].question;
        this.checkSurveyInteraction();
        this.changeDetector.detectChanges();
      }
    });
  }

  private checkSurveyInteraction(): void {
    const hasInteractedWithSurvey = localStorage.getItem(`hasInteractedWithSurvey_${this.surveyID}`);
    this.showSurvey = !hasInteractedWithSurvey;
    this.changeDetector.detectChanges();
  }

  handleDismiss(): void {
    this.showSurvey = false;

>     localStorage.setItem(`hasInteractedWithSurvey_${this.surveyID}`, 'true');

Reviewer: Should we use the same localStorage format that PostHog does?

It looks like `seenSurvey_${this.surveyID}` 

Don't feel strongly.

Author: Dont think it really matters tbh. This string can be anything, as long as it's unique

        const survey = surveys[0];
        this.surveyID = survey.id;
        this.surveyTitle = survey.questions[0].question;
        this.checkSurveyInteraction();
        this.changeDetector.detectChanges();
      }
    });
  }

  private checkSurveyInteraction(): void {
    const hasInteractedWithSurvey = localStorage.getItem(`hasInteractedWithSurvey_${this.surveyID}`);
    this.showSurvey = !hasInteractedWithSurvey;
    this.changeDetector.detectChanges();
    if (this.showSurvey) {
      posthog.capture("survey seen", {
        $survey_id: this.surveyID // required
      })
    }
  }

  handleDismiss(): void {
    this.showSurvey = false;
    localStorage.setItem(`hasInteractedWithSurvey_${this.surveyID}`, 'true');
    posthog.capture("survey dismissed", {
      $survey_id: this.surveyID // required
    })
  }

  handleSubmit(value: number): void {
    console.log("Submitted value:", value);
    this.showSurvey = false;
    localStorage.setItem(`hasInteractedWithSurvey_${this.surveyID}`, 'true');
    posthog.capture("survey sent", {
      $survey_id: this.surveyID, // required
      $survey_response: `${value}` // required. Convert numbers to string
    })
  }
}
```

> Our survey is now ready to go! The next step is ship the changes, get responses, and view your results.

Reviewer: ```suggestion
Our survey is now ready to go! The next step is to ship the changes, get responses, and view your results.
```
